### PR TITLE
Verify Parent Instruction Execution

### PR DESCRIPTION
**Pull request by talekhinezh**
* Implement `VERIFY_PARENT` by creating a temporary child AuthZone to the parent Authzone and checking an access rule against this AuthZone
* Remove unwraps from `MultithreadIntentProcessor`

### CODE REVIEW

**Review by iamyulong — APPROVED** [review]

File: radix-engine/src/system/transaction/multithread_intent_processor.rs [hunk 1]
```diff
@@ -138,6 +196,47 @@ impl MultiThreadIntentProcessor {
         Ok(())
     }
 
+    fn create_temp_auth_zone<Y: SystemBasedKernelApi>(api: &mut Y) -> Result<NodeId, RuntimeError> {
```
Comment (dhedey) [inline radix-engine/src/system/transaction/multithread_intent_processor.rs:199]: MINOR: If we keep this, from a code clarity perspective, I'd ideally like to explain what this does more clearly, e.g.

```rust
fn create_child_auth_zone_with_transaction_processor_global_caller(...)
```

File: radix-engine/src/system/transaction/multithread_intent_processor.rs [hunk 2]
```diff
@@ -92,44 +98,96 @@ impl MultiThreadIntentProcessor {
...
 
-            // TODO: Remove unwraps
             let mut system_service = SystemService::new(api);
-            let switch_to = match txn_thread.resume(passed_value.take(), &mut system_service)? {
+            let post_exec = match txn_thread.resume(passed_value.take(), &mut system_service)? {
                 ResumeResult::YieldToChild(child, value) => {
-                    let child = *children_mapping.get(child).unwrap();
+                    let child = *children_mapping
+                        .get(child)
+                        .ok_or(RuntimeError::SystemError(SystemError::IntentError(
```
Comment (dhedey) [inline radix-engine/src/system/transaction/multithread_intent_processor.rs:116]: Nice tweaks 👍.

File: radix-engine-tests/tests/system/subintent_verify_parent.rs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    );
+
+    let receipt = ledger.execute_test_transaction(transaction);
+
+    // Assert
+    receipt.expect_commit_success();
+}
+
+#[test]
+fn should_be_able_to_use_subintent_when_verify_parent_access_rule_is_met_on_second_yield() {
```
Comment (dhedey) [inline radix-engine-tests/tests/system/subintent_verify_parent.rs:81]: Interesting, nice test 👍  (and nice tests :D )

**Review by dhedey — COMMENTED** [review]
Nice! A few thoughts we should quickly chat through

File: radix-engine/src/system/transaction/multithread_intent_processor.rs (hunk 1 above)
Comment (dhedey) [reply radix-engine/src/system/transaction/multithread_intent_processor.rs:199]: Okay, after having it explained, perhaps we could call it:
```rust
fn create_temporary_child_auth_zone_for_verify_parent(api: impl SystemBasedKernelApi) -> Result<NodeId, RuntimeError> {
//..
}
```

File: radix-engine/src/system/transaction/multithread_intent_processor.rs [hunk 4]
```diff
@@ -92,44 +98,96 @@ impl MultiThreadIntentProcessor {
...
+                            .ok_or(RuntimeError::SystemError(SystemError::IntentError(
+                                IntentError::CannotVerifyParentOnRoot,
+                            )))?;
+                    api.kernel_switch_stack(parent)?;
 
-                cur_thread = next_thread;
-            } else {
-                Self::cleanup_stack(api)?;
-                break;
+                    // Create a temporary authzone with the current authzone as the global caller
```
[the remark this answers is not in the document]
Comment (dhedey) [reply radix-engine/src/system/transaction/multithread_intent_processor.rs:164]: After clarification with you, it sounds like the global caller is irrelevant?

But perhaps in this comment we could briefly explain that an access rule assertion is against the ancestor call frames, so we need to create a child call frame to assert against the content of the auth zone?

[the remark this answers is not in the document]
Comment (dhedey) [reply radix-engine/src/system/transaction/multithread_intent_processor.rs:164]: EDIT: After discussion, I understand now 👍  Will mark this as resolved.

File: radix-engine/src/system/transaction/instructions.rs [hunk 5]
```diff
@@ -186,7 +196,21 @@ impl YieldingInstruction for YieldToParent {
...
+}
+
+impl MultiThreadInstruction for VerifyParent {
+    fn execute<Y: SystemApi<RuntimeError> + KernelNodeApi + KernelSubstateApi<L>, L: Default>(
+        self,
+        _worktop: &mut Worktop,
+        _objects: &mut IntentProcessorObjects,
+        _api: &mut Y,
+    ) -> Result<MultiThreadResult, RuntimeError> {
+        // TODO: Remove unwrap
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/system/transaction/instructions.rs:210]: sweet :)

File: radix-engine/src/system/transaction/multithread_intent_processor.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/system/transaction/multithread_intent_processor.rs:164]: Yup it's for the most part irrelevant but probably what makes most sense I think. Will add comment

**Pull request merged**